### ISSUE TITLE
Batching results in data loss

### ISSUE THREAD

**Issue by trevhunter** [influxdata/influxdb-csharp#46]
**Expected:** I should be able to use batching and not lose points. 
**Actual:**  If I use batching, then data is lost when sending lots of points in quick succession.
**Version:** Using V1.0.0 of InfluxDB.Collector and V1.0.0 of InfluxDb.LineProtocol from Nuget

 [See attached file for recreation](https://github.com/influxdata/influxdb-csharp/files/1589796/InfluxCollectorBug.zip). It includes the following code:

``` csharp
static void Main(string[] args)
{

    WriteLotsOfMetricsFast("UnbatchedPoints", useBatching: false); // Results in all 1000 points being written to InfluxDb
    WriteLotsOfMetricsFast("BatchedPoints", useBatching: true); // Results in very few points being written (about 3 on my machine);

    Console.WriteLine("Waiting for 5s to flush the batched stuff");
    System.Threading.Thread.Sleep(5000);

}


private static void WriteLotsOfMetricsFast(string measureName, bool useBatching)
{
    CollectorConfiguration config;
    if (useBatching)
    {
        config = new CollectorConfiguration().Batch.AtInterval(TimeSpan.FromSeconds(1));
    }
    else
    {
        config = new CollectorConfiguration();
    }
    config = config.WriteTo.InfluxDB("http://localhost:8086", "MyData");
    Metrics.Collector = config.CreateCollector();


    for (int i = 0; i < 1000; i++)
    {
        var fields = new Dictionary<string, object>()
        {
                 { "Iteration", i },
                 { "MyValue", 10}
         };

        Metrics.Write(measureName, fields, null);
    }
}
```

**Comment by trevhunter** [comment]
Further investigation reveals this seems to be down to the resolution of `DateTime.UtcNow` which is usually in the region of 100ns. MetricsCollector.cs uses `DateTime.UtcNow` to specify the timestamp and if points are created in quick succession (i.e. < 100ns) then Influx will discard them.

Given this, it seems that this is related to #34 which requests that nanosecond precision be available.

**Comment by gambrose** [comment]
I don't think it would be a precision issue. DateTime is 100ns precision but it is only updated about very every 10ms on context switch.

**Comment by trevhunter** [comment]
I'm not really worried about the precision here, just the potential for unintentional information loss by calling Write too fast.

Where this might actually show up in reality, rather than in the contrived loop that I created in the example is if some code submits 2 or more points in quick succession to the same measure.

Using DateTime.UtcNow as the default timestamp could result in some of the points being dropped.

I'm not sure of a good remedy, but it could be one of:

1) Auto increment DateTime.UtcNow by 1 tick if Write is called within a timescale that results in the same previous DateTime.UtcNow.Ticks value. Only do this for non-user supplied timescales. 

2) Use high precision timer. Probably not available on all platforms.

**Comment by gambrose** [comment]
I've never used the Metrics package, I've only been involved in the line protocol writing code, so I cannot speak with any authority on what the intention of the Metrics code is trying to do. I also am struggling to understand your use case so please do not take any of the below as patronising.

It seams to me that basing the sampling period on `DateTime.UtcNow` is the correct thing to do. It has very little overhead and will fit 99.9% of use cases.
If you need the extra precision you can supply your own timestamp which can be taken from a higher precision timer such as a `Stopwatch` that you are maintaining.

The behavior you are describing as points going missing is in fact points being updated to the latest value and is a core influx design concept of last write wins (same timestamp same point). But it is important to understand this only happens then the measurement name, timestamp, and **tag set** match. I.e. if you where writing 1000 points with different tags all the points would be kept as they would all be unique.

It may help if you could expand on why you do not want points to be overwritten in a series. I can only think of a few use cases where this would be an issue;
1) Your problem domain is high precision, like monitoring wave forms in electronics, in which case I would assume you would be supplying your own timestamps as timing is very important.
2) You care about the total number of times something happened. In which case I would recommend incrementing an (interlocked) counter and then recording the value of the counter as a metric value. This is how a lot of metrics are captured when using Prometheus (the tool I currently use). It allows for the system state to be understood even if some of the points maybe lost. Which for metrics is a nice attribute as I'm mostly interested when things are failing. You would likely have a separate counter for each thing you are interested in, passes, fails.

**Comment by trevhunter** [comment]
@gambrose You're not being patronizing at all - I'm hoping I can help here too as its a fantastic library. You're absolutely right about the root cause being expected behavior on "last write wins". That said, the underlying behavior doesn't match my expectations for the `Collector` API and would prevent me from using it directly in an application without some form of wrapper. 

Let me attempt a better explanation with a more realistic example likely to happen. I'd consider it a common example, yet it gets 80% success, not 99.9%.

**Problem**
I have a web server that processes authentication requests. I use the InfluxDB.Collector.Metrics API to log successful and unsuccessful attempts to authenticate. The web server has multiple cores and can process multiple requests in parallel. Influx is set up to use nanosecond precision, but I don't really need better than 1s precision when querying the data. The call I use to log a successful and unsuccessful attempt is like the below:

```csharp
var tags = var tags = new Dictionary<string, string>()
                {
                    { "Status" ,  (isSuccess ? "Success" : "Failure") }
                };

Metrics.Increment("AuthenticateAttempt", 1, tags};
```
Running the above with only modest parallelism (8 threads with 250ms for each login attempt), I get a 20% data loss (i.e. 20% of the time, there's a clash in `DateTime.UtcNow` even when there's only 8 threads processing things at 250ms spacing of each attempt). 


**Analysis**
I understand the cause, but the 20% data loss really surprised me. The 100ns resolution of `DateTime.UtcNow` really adds up quickly even under very modest load.

(Yes, I realize the prod system would have more than one server, and collisions on the timestamp are very possible across machines - easily avoided by adding "host" to the tags.)

This isn't just related to "Count" (although that's what I'm illustrating here). The low resolution timestamp combined with multiple threads would mean any type of data gets overwritten by Influx.

This is a hard problem to solve because it's all about expectations - it's expected behavior at the InfluxDB level because we're supplying it with low resolution time-stamps. What's not expected is that these timestamps collide more often than one might think. And for the higher level Collector API with methods like `Increment`, users should be more acutely aware of its limitations.

**Possible Solutions**
Assuming the InfluxDB.Collector API is intended to be high level, this means that it or the consumer of the API needs to wrap it and do one of the following:

(a) Specify custom timestamps. Even if a user specified their own timestamps, they're likely to also use DateTime.UtcNow - since that's the only thing most application developers are aware of. 

(b) Add some kind of tag - maybe thread id or sequence number or something so Influx won't collapse the points. Seems wasteful in the data. 

(c) Do their own batching - cache writes in memory, increment them appropriately to cover the DateTime.UtcNow resolution, then send them through the api. Probably non-trivial and something that statsd and such are for.

(d) Use some form of higher resolution Tick source than DateTime.UtcNow (perhaps `System.Diagnostics.Stopwatch.GetTimeStamp()`). Probably still subject to the same issue if the influx precision is set to something lower.

(e) Manually fudge the DateTime.UtcNow.Ticks value by incrementing it. Also probably won't fix the issue if the underlying precision is lower.

To me, (c) is probably the most correct for the "Collector" API, but likely very non-trivial to implement. (b)  is a quick win, but messy. (a) (d) (e) don't really solve the problem but (e) would likely raise the success rate from 80% to over 99.9% in the common use case.

**Comment by gambrose** [comment]
@trevhunter thanks for clarifying I was unaware of the `Metrics.Increment` and I would agree with you that I would not expect this method to lose increments.  I'm afraid I have rather derailed this issue really @nblumhardt is the guy you need to be talking to as as far as I understand the Api is creation.

If you are trying to log high frequency events and do not want to miss any then would urge you to consider using counters that last the lifetime of your application rather than try and log each event separately.  It solves the issue in short term `DateTime.UtcNow` but also if the server goes down for say a minute due to network outage. You will not have lost any points as the total would be correctly reported when the network recovers, you would not know when during the min they occurred but you would at least know they did occur.
It requires extra work on reporting as you need to graph the "rate" of increase rather than just the "sum" but it also has the benefit of positive reassurance if a value is not changing. I.e. is this value really not increasing or are my points just getting not getting logged, a total asserts that there has been no changes since last logged value.
Influx includes the `NON_NEGATIVE_DIFFERENCE` function to report on data recorded in this way. The non negative bit allows you to sum increments across app restarts.

**Comment by optical** [comment]
I've personally encountered this issue in a different InfluxDB library, and I definitely think its one that should be addressed internally without any API changes. As already mentioned, methods like Increment() and even Time() called in rapid succession should never lose data.

The proposal to increment the ns of the timestamp by one, if its the same timestamp previously recorded seems like the most robust solution, and is the approach I've seen used elsewhere. It won't require any API changes, and as far I am aware is fairly bullet proof in any real world scenario, except if you're somehow able to write stats at a rate > 1stat/ns.

Not to dismiss any of the other proposals here. I absolutely agree that consumers shouldn't spam influx with lots of stats, batching, aggregation, intelligent tagging and internally incremented counters are all solid advice in general, but even when following them you can still inadvertently get stung by this race condition.

**Comment by nblumhardt** [comment]
👍 implementing (e) at the collector level seems like the option most likely to be achievable in the short term. Is anyone interested in having a shot at it? My time for hacking at this is a bit short currently :-)

**Comment by trevhunter** [comment]
@nblumhardt I'll take it on and get a PR to you for review shortly. I don't know if this is the right solution for this api long term, but I'll open a pull request anyway and the maintainers can decide what to do with it.  

I've implemented a fairly simple `PseudoHighResolutionTimestampSource` and am using it in my wrapper API with pretty good success. It's a little brute force, but it solves for the "out of the box" configuration nicely (where out of the box = InfluxDb nanosecond precision... it would start to fail again if the user started specifying a different precision, but at that point, I assume they know what they're doing.)
